fendermint/vm/interpreter/src/chain.rs

```diff
@@ -221,32 +238,51 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDownProposal(_) => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownProposal(p) => {
+                    let validator_set = downcast_err!(
+                        atomically_or_err(|| provider.validator_set(p.height as u64)).await
+                    )?
+                    .ok_or_else(|| anyhow!("cannot find validator set for block: {}", p.height))?;
+
+                    let finality = IPCParentFinality {
+                        height: p.height as u64,
+                        block_hash: p.block_hash,
+                    };
+                    let msg = parent_finality_to_fvm(finality.clone(), validator_set)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+
+                    downcast_err!(
+                        atomically_or_err(|| provider.on_finality_committed(&finality)).await
+                    )?;
+
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
```

Comment:
Can you point me where we send the message to commit the parent finality?

Reply:
discussed this over slack.